src/import_vdf/vertexai_vector_search_import.py

```diff
@@ -0,0 +1,861 @@
+            )
+
+        except Exception as e:
+            print(
+                f"Failed to deploy {self.index_name} to endpoint: {self.index_endpoint_name}"
+            )
+            raise e
+
+        return index_endpoint
+
+    def upsert_data(self):
+        MINUTE = 60
+        CALLS_PER_PRD = self.args.get("requests_per_minute", 6000)
+
+        # rate is 1 QPS
+        @sleep_and_retry  # If there are more requests than rate, sleep shortly
+        @on_exception(
+            expo, google_exceptions.ResourceExhausted, max_tries=10
+        )  # if we receive exceptions from Google API, retry
+        @limits(calls=CALLS_PER_PRD, period=MINUTE)
+        def upsert_in_rate(self, upsert_request):
+            """
+            Upserts vectors and applies rate limits.
+            """
+            return self.index_client.upsert_datapoints(request=upsert_request)
+
+        # upsert for each index
+        for new_index_name, (index_name, index_meta) in tqdm(zip(
+            self.index_names, self.vdf_meta["indexes"].items()
+        ), desc="Iterating over indexes"):
+            self.index_name = new_index_name
+            self.target_index = self._get_index()
+            self.target_index_resource_name = self.target_index.name
+            # init target index to import vectors to
+            self.target_vertexai_index = aip.MatchingEngineIndex(
+                self.target_index_resource_name
+            )
+            # load data
+            tqdm.write(
+                f"Importing data from {index_name} to {self.target_vertexai_index.display_name}"
+            )
+            tqdm.write(f"index_meta: {json.dumps(index_meta, indent=4)}")
+
+            for namespace_meta in tqdm(index_meta, desc="Iterating over namespaces"):
+                # get data path
+                data_path = namespace_meta["data_path"]
+                final_data_path = self.get_final_data_path(data_path)
+                print(f"data_path: {data_path}")
+
+                # get col names
+                vector_metadata_names, vector_column_name = self.get_vector_column_name(
+                    namespace_meta["vector_columns"], namespace_meta
+                )
+                print(f"vector_column_name    : {vector_column_name}")
+                print(f"vector_metadata_names : {vector_metadata_names}")
+
+                # Load the data from the parquet files
+                parquet_files = self.get_parquet_files(final_data_path)
+
+                total_ids = []
+                for file in tqdm(parquet_files, desc="Iterating over parquet files"):
+                    file_path = os.path.join(final_data_path, file)
+                    df = pd.read_parquet(file_path)
+                    df["id"] = df["id"].apply(lambda x: str(x))
+
+                    insert_datapoints_payload = []
+
+                    for idx, row in tqdm(
+                        df.iterrows(), desc="Iterating over rows", total=len(df)
+                    ):
```

Comment:
this loop can be rewritten to be more concise and efficient. Will leave it for you to do in a subsequent PR.